Log public key in addition to IP address

a1q123456 opened the pull request:
## High Level Overview of Change

This PR adds public keys to logs.

### Context of Change

Some validators used to complain that we're logging ip addresses which can expose them to attacks. However, after we masked out the ip addresses from the logs, we're unable to debug some issues. This PR adds public key to logs to help us debug issues. 

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

There's no API impact.

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

## Test Plan
Regression test.

a1q123456 commented:
Example
```
2025-Aug-13 17:26:58.331162 UTC Resource:TRC Charging IP Address: x.x.x.x:51235, Public Key: 02F21D04A551A2CD1039 for trivial peer request ($1) (tx)

```

Tapanito commented on include/xrpl/protocol/PublicKey.h:
```diff
@@ -128,8 +129,28 @@ class PublicKey
+    Slice
+    shortId() const noexcept
+    {
+        return slice().substr(0, 10);
```
I noticed that Node IDs are typically represented using base58 encoding, generated by `toBase58(TokenType::NodePublic, ...)`. What do you think about following that convention here to maintain consistency?

a1q123456 commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -609,8 +608,8 @@ PeerImp::fail(std::string const& reason)
         std::string const n = name();
-        JLOG(journal_.warn()) << (n.empty() ? remote_address_.to_string() : n)
-                              << " failed: " << reason;
+        JLOG(journal_.warn())
```
Thanks, Vito! That's really helpful!

Tapanito commented on src/xrpld/overlay/detail/PeerImp.cpp:
There are two places, in `fail` and `onAccept` where we already output the encoded public key. Could you remove them please?

Tapanito reviewed (approved):
LGTM

a1q123456 commented on src/libxrpl/resource/Consumer.cpp:
```diff
@@ -148,6 +148,12 @@ Consumer::entry()
 }
 
+void
+Consumer::setPublicKey(PublicKey const& publicKey)
```
I think this is because of optimisation

bthomee commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -730,6 +726,16 @@ PeerImp::setTimer(std::chrono::seconds interval)
+PeerImp::makePrefix(std::string const& fingerprint)
+{
+    std::stringstream ss;
+    ss << "[" << std::setfill('0') << std::setw(3) << fingerprint << "] ";
```
The `std::setw(3)` seems unnecessary, since the updated fingerprint now contains the IP address, public key, and node ID - which will be much longer than 3 characters.

bthomee commented on src/xrpld/peerfinder/detail/Logic.h:
```diff
@@ -360,7 +360,7 @@ class Logic
     {
         JLOG(m_journal.trace())
-            << beast::leftw(18) << "Logic connected" << slot->remote_endpoint()
+            << beast::leftw(18) << "Logic connected" << slot->fingerprint()
```
```suggestion
            << beast::leftw(18) << "Logic connected " << slot->fingerprint()
```

I think an extra space is necessary after `Logic connected` to separate the fingerprint.

bthomee commented on src/xrpld/peerfinder/detail/Logic.h:
```diff
@@ -968,12 +968,12 @@ class Logic
             case Slot::closing:
                 JLOG(m_journal.trace()) << beast::leftw(18) << "Logic finished "
-                                        << slot->remote_endpoint();
+                                        << slot->fingerprint();
```
I'm noticing a general inconsistency in how the logging happens. In PeerImp the sinks were modified to use the fingerprint as prefix, but here the sinks are unmodified but the fingerprint is added explicitly to the log message.

I would suggest to revisit the codebase add the fingerprint as prefix to sinks wherever possible. If it isn't possible in most places, so you'll have to add it manually to the log lines for many sinks, then maybe it is better to not use the fingerprint as sink prefix anywhere and just add to each log line. I'm otherwise concerned the logs end up in a hard-to-parse mess.

a1q123456 replied:
Tried running rippled for around 10 minutes for each branch and the result shows that the develop branch writes around 50 MB logs per minute and this PR writes 66 MB logs per minute, so I'd say this PR is going to increase the log file size by around 30% - 32%.

Version | Duration (min) | Growth rate (bytes/min) | Growth rate (MB/min)
-- | -- | -- | --
PR | 10.888 | 66,739,865 | 66.74 MB/min
develop | 10.727 | 50,785,364 | 50.79 MB/min

bthomee reviewed (commented):
Due to adding the fingerprint as a sink prefix, the logs will be growing as in the past only IDs were logged and now IP address and public key as well. By how much do you expect the logs to grow?

bthomee commented on src/xrpld/peerfinder/detail/SlotImp.h:
```diff
@@ -91,6 +91,12 @@ class SlotImp : public Slot
     }
 
+    std::string
+    logPrefix() const
```
For the peer, you named this function just `prefix()` but here it is named `logPrefix()` while they both serve the same purpose - wrapping the fingerprint in brackets for printing in a log message.

Can you name both functions the same?

bthomee commented on src/xrpld/peerfinder/detail/Logic.h:
```diff
@@ -682,8 +686,8 @@ class Logic
-                    << beast::leftw(18) << "Logic sending "
-                    << slot->remote_endpoint() << " with " << list.size()
+                    << beast::leftw(18) << "Logic sending " << slot->logPrefix()
+                    << " with " << list.size()
```
Why are you using `JLOG(m_journal....)` here with the fingerprint inline, instead of using

```
beast::WrappedSink sink{m_journal.sink(), slot->logPrefix()};
beast::Journal journal{sink};
JLOG(journal...)
```
?

I'd like the logs to provide a consistent experience, rather than having the fingerprint appear at the beginning in some log lines, while in others it's in the middle or at the end.

a1q123456 replied:
Was just thinking `journal` would only need to be used once and it was probably not worth it.
But yeah, that makes sense. I've updated the code.

bthomee commented on src/xrpld/peerfinder/detail/Logic.h:
```diff
@@ -835,7 +842,7 @@ class Logic
                 {
-                    JLOG(m_journal.debug())
+                    JLOG(journal.debug())
                         << beast::leftw(18) << "Logic testing " << ep.address
```
Should the `ep.address` remain here? More specifically, it the address of the node itself, which is already part of the fingerprint that will be printed by the journal, or is it the address of its peer and it thus should stay?

a1q123456 replied:
It should, because `ep` is from peer protocol and it's not going to be the same as the remote address of the slot.

bthomee commented on src/xrpld/peerfinder/detail/Logic.h:
```diff
@@ -788,16 +793,18 @@ class Logic
-            << beast::leftw(18) << "Endpoints from " << slot->remote_endpoint()
-            << " contained " << list.size()
+        JLOG(journal.trace())
+            << beast::leftw(18) << "Endpoints contained " << list.size()
```
All these `leftw(18)` statements here and elsewhere probably can go, now you've modified the journals to print many more characters than before.

a1q123456 replied:
Thanks, Bart! I've removed them.

bthomee committed:
feat: Add public key to log messages

To protect the identity of UNL validators, the IP addresses are redacted from the log messages sent to the common Grafana instance. However, without such identifying information it is challenging to debug issues. This change adds a node's public key to logs to improve our ability to debug issues.
